src/struct_ls/HYPRE_struct_pcg.c

```diff
@@ -210,43 +210,49 @@ HYPRE_StructDiagScale( HYPRE_StructSolver solver,
    HYPRE_Real           *xp;
 
    hypre_Index           index;
-   hypre_IndexRef        start;
-   hypre_Index           stride;
+
+   HYPRE_Int             nboxes;
+   hypre_Box            *loop_box;
    hypre_Index           loop_size;
+   hypre_IndexRef        start;
+   hypre_Index           ustride;
 
    HYPRE_Int             i;
 
-   /* x = D^{-1} y */
-   hypre_SetIndex(stride, 1);
-   boxes = hypre_StructGridBoxes(hypre_StructMatrixGrid(A));
-   hypre_ForBoxI(i, boxes)
+   nboxes = hypre_StructVectorNBoxes(x);
+
+   loop_box = hypre_BoxCreate(ndim);
+   hypre_SetIndex(ustride, 1);
+
+   for (i = 0; i < nboxes; i++)
    {
-      box = hypre_BoxArrayBox(boxes, i);
+      hypre_StructVectorGridBoxCopy(x, i, loop_box);
+      start = hypre_BoxIMin(loop_box);
 
       A_data_box = hypre_BoxArrayBox(hypre_StructMatrixDataSpace(A), i);
```

Comment:
I'm wondering if we should have a `hypre_StructMatrixGridDataBox` as we have for vectors now